cmd/xl-storage-format-v2.go

```diff
@@ -0,0 +1,437 @@
+	Type         JournalType           `json:"type" msg:"type"`
+	DeleteMarker *XLMetaV2DeleteMarker `json:"delete,omitempty" msg:"delete,omitempty"`
+	Object       *XLMetaV2Object       `json:"object,omitempty" msg:"object,omitempty"`
+}
+
+// Valid xl meta XLMetaV2JournalEntry is valid
+func (j XLMetaV2JournalEntry) Valid() bool {
+	switch j.Type {
+	case ObjectType:
+		if j.Object == nil {
+			return false
+		}
+		return isXLMetaErasureInfoValid(j.Object.DataErasureN, j.Object.DataErasureM)
+	case DeleteType:
+		if j.DeleteMarker == nil {
+			return false
+		}
+		if j.DeleteMarker.ModTime <= 0 {
+			return false
+		}
+		return true
+	}
+	return false
+}
+
+// XLMetaV2 - object meta structure defines the format and list of
+// the journals for the object.
+type XLMetaV2 struct {
+	ObjectJournals []XLMetaV2JournalEntry `json:"ojs" msg:"ojs"`
+}
+
+// AddVersion adds a new version
+func (z *XLMetaV2) AddVersion(fi FileInfo) error {
+	var uv uuid.UUID
+	var err error
+	if fi.VersionID != nullVersionID && fi.VersionID != "" {
+		uv, err = uuid.Parse(fi.VersionID)
+		if err != nil {
+			return err
+		}
+	}
+
+	dd, err := uuid.Parse(fi.DataDir)
+	if err != nil {
+		return err
+	}
+
+	jentry := XLMetaV2JournalEntry{
+		Type: ObjectType,
+		Object: &XLMetaV2Object{
+			VersionID:               uv,
+			DataDir:                 dd,
+			StatSize:                fi.Size,
+			StatModTime:             fi.ModTime.Unix(),
+			DataErasureAlgorithm:    ReedSolomon,
+			DataErasureM:            fi.Erasure.DataBlocks,
+			DataErasureN:            fi.Erasure.ParityBlocks,
+			DataErasureBlockSize:    fi.Erasure.BlockSize,
+			DataErasureIndex:        fi.Erasure.Index,
+			DataErasureChecksumAlgo: HighwayHash,
+			DataErasureDistribution: make([]uint8, len(fi.Erasure.Distribution)),
+			DataPartInfoNumbers:     make([]int, len(fi.Parts)),
+			DataPartInfoETags:       make([]string, len(fi.Parts)),
+			DataPartInfoSizes:       make([]int64, len(fi.Parts)),
+			DataPartInfoActualSizes: make([]int64, len(fi.Parts)),
+			MetaSys:                 make(map[string][]byte),
+			MetaUser:                make(map[string]string),
+		},
+	}
+
+	for i := range fi.Erasure.Distribution {
+		jentry.Object.DataErasureDistribution[i] = uint8(fi.Erasure.Distribution[i])
+	}
+
+	for i := range fi.Parts {
+		jentry.Object.DataPartInfoSizes[i] = fi.Parts[i].Size
+		if fi.Parts[i].ETag != "" {
+			jentry.Object.DataPartInfoETags[i] = fi.Parts[i].ETag
+		}
+		jentry.Object.DataPartInfoNumbers[i] = fi.Parts[i].Number
+		jentry.Object.DataPartInfoActualSizes[i] = fi.Parts[i].ActualSize
+	}
+
+	for k, v := range fi.Metadata {
+		if strings.HasPrefix(k, ReservedMetadataPrefix) || strings.HasPrefix(k, strings.ToLower(ReservedMetadataPrefix)) {
+			jentry.Object.MetaSys[k] = []byte(v)
+		} else {
+			jentry.Object.MetaUser[k] = v
+		}
+	}
+	z.ObjectJournals = append(z.ObjectJournals, jentry)
+	return nil
+}
+
+func newXLMetaV2(fi FileInfo) (XLMetaV2, error) {
+	xlMeta := XLMetaV2{}
```

Comment:
I may be missing something, but could this function be replaced by:

```suggestion
	xlMeta := XLMetaV2{}
	return xlMeta, xlMeta.AddVersion(fi)
```